--- ./op.c.orig	Mon Aug  4 14:57:00 2003
+++ op.c	Mon Aug  4 14:57:30 2003
@@ -1561,6 +1561,7 @@ S_apply_attrs_my(pTHX_ HV *stash, OP *ta
     (void)SvUPGRADE(meth, SVt_PVIV);
     (void)SvIOK_on(meth);
     PERL_HASH(SvUVX(meth), SvPVX(meth), SvCUR(meth));
+    pack->op_private |= OPpCONST_BARE; /* for autobox */
     imop = convert(OP_ENTERSUB, OPf_STACKED|OPf_SPECIAL|OPf_WANT_VOID,
 		   append_elem(OP_LIST,
 			       prepend_elem(OP_LIST, pack, list(arg)),
@@ -2974,6 +2975,7 @@ Perl_utilize(pTHX_ int aver, I32 floor, 
 	(void)SvUPGRADE(meth, SVt_PVIV);
 	(void)SvIOK_on(meth);
 	PERL_HASH(SvUVX(meth), SvPVX(meth), SvCUR(meth));
+	pack->op_private |= OPpCONST_BARE; /* for autobox */
 	imop = convert(OP_ENTERSUB, OPf_STACKED|OPf_SPECIAL,
 		       append_elem(OP_LIST,
 				   prepend_elem(OP_LIST, pack, list(arg)),
@@ -6074,6 +6076,28 @@ Perl_ck_subr(pTHX_ OP *o)
     else if (cvop->op_type == OP_METHOD || cvop->op_type == OP_METHOD_NAMED) {
 	if (o2->op_type == OP_CONST)
 	    o2->op_private &= ~OPpCONST_STRICT;
+
+	/*
+	   disable autoboxing if Name in "Name->method()" is a bareword
+	   otherwise, enable it if the appropriate field of the
+	   (lexically-scoped) hints hash is set
+	 */
+	if (!(o2->op_private & OPpCONST_BARE)
+		&& (PL_hints & HINT_LOCALIZE_HH)
+		&& (cvop->op_type == OP_METHOD_NAMED)) {
+		HV *table = GvHV(PL_hintgv);
+		SV **svp;
+
+		if (table && (svp = hv_fetch(table, "AUTOBOX", 7, FALSE))
+		    && *svp && SvOK(*svp)) {
+		    SV * new_sv = *svp;
+		    SV * old_sv = ((SVOP *)cvop)->op_sv;
+
+		    sv_catpvn(old_sv, "\0", 1);
+		    sv_catsv(old_sv, new_sv);
+		    cvop->op_flags |= OPf_SPECIAL;
+		}
+	}
 	else if (o2->op_type == OP_LIST) {
 	    OP *o = ((UNOP*)o2)->op_first->op_sibling;
 	    if (o && o->op_type == OP_CONST)
