--- ./pp_hot.c.orig	Mon Aug  4 14:57:20 2003
+++ pp_hot.c	Mon Aug  4 14:57:40 2003
@@ -3036,7 +3036,7 @@ STATIC SV *
 S_method_common(pTHX_ SV* meth, U32* hashp)
 {
     SV* sv;
-    SV* ob;
+    SV* ob = NULL;
     GV* gv;
     HV* stash;
     char* name;
@@ -3053,6 +3053,58 @@ S_method_common(pTHX_ SV* meth, U32* has
 
     if (SvGMAGICAL(sv))
 	mg_get(sv);
+
+    if ((PL_op->op_flags & OPf_SPECIAL) && (!SvOBJECT(SvROK(sv) ? (ob = SvRV(sv)) : sv))) {
+	/* set ob in case we goto fetch */
+	HV * autobox_cache; /* points to the primitive2boxed map */
+	char *packed = SvPV(meth, SvCUR(meth)); /* augmented (hacked) method name */
+	STRLEN klen; /* length of the original method name */
+
+	klen = strlen(packed) + 1; /* up to "\0" */
+	/* meth = newSVpvn(packed, klen); */
+	meth = sv_2mortal(newSVpvn(packed, klen)); /* original method name */
+	name = SvPV(meth, namelen); /* update name and namelen for diagnostics */
+	PERL_HASH(*hashp, name, namelen); /* update (i.e. correct) the hash code */
+	autobox_cache = get_hv("autobox::cache", 0);
+
+	if (autobox_cache) {
+	    SV * autobox_key; /* %autobox::cache key */
+	    HV * autobox_handlers = NULL; /* %autobox::cache value */
+	    HE * he; /* %autobox::cache entry */
+	    SV **svp; /* pointer to autobox_handlers */
+	    char *reftype; /* autobox_handlers key */
+
+	    autobox_key = newSVpv(packed + klen, 0);
+	    he = hv_fetch_ent(autobox_cache, autobox_key, 0, 0);
+	    
+	    if (he) {
+		SV *heval = HeVAL(he);
+		if (SvROK(heval) && (autobox_handlers = (HV *)(SvRV(heval)))) {
+		    reftype = sv_reftype((ob ? ob : sv), 0);
+		    svp = hv_fetch(autobox_handlers, reftype, strlen(reftype), FALSE);
+
+		    if (svp) {
+			if (SvOK(*svp)) { /* i.e. sv is not undef */
+			    packname = SvPVX(*svp); /* fake the package name */
+			    packlen = strlen(packname);
+			    stash = gv_stashpvn(packname, packlen, FALSE);
+			    /* set packsv in case we goto fetch */
+			    packsv = *svp;
+
+			    if (stash) {
+				SV* ref = newSViv(PTR2IV(stash));
+				/* cache the stash */
+				hv_store(PL_stashcache, packname, packlen, ref, 0);
+			    }
+			    /* either way goto fetch, as we've *demanded* autoboxing */
+			    goto fetch;
+			} /* else: autobox_handlers value is undef - don't autobox */
+		    } /* else autobox_handlers key doesn't exist - don't autobox */
+		} /* else %autobox::cache value is not a hash ref - don't autobox */
+	    } /* else: %autobox::cache key doesn't exist - don't autobox */
+	}
+    }
+
     if (SvROK(sv))
 	ob = (SV*)SvRV(sv);
     else {
